<a href="https://colab.research.google.com/github/Deyonrose/Speech_Processing/blob/main/2348513_Lab8_SPR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**implement the Viterbi Algorithm**

The Viterbi Algorithm is a computational technique designed to determine the most likely sequence of hidden states in a Hidden Markov Model (HMM) based on observed data. By systematically evaluating transition and emission probabilities for each time step, the algorithm identifies the optimal sequence of states through a process of dynamic programming and backtracking. This makes it a powerful tool for decoding sequences in fields such as speech recognition and bioinformatics.

In [1]:
import numpy as np

def viterbi_algorithm(observation_sequence, states, start_prob, trans_prob, emis_prob):
    n_states = len(states)
    n_observations = len(observation_sequence)

    # Initialize the viterbi matrix and backpointer matrix
    viterbi = np.zeros((n_states, n_observations))
    backpointer = np.zeros((n_states, n_observations), dtype=int)

    # Initialization step
    for s in range(n_states):
        viterbi[s, 0] = start_prob[s] * emis_prob[s, observation_sequence[0]]
        backpointer[s, 0] = 0

    # Recursion step
    for t in range(1, n_observations):
        for s in range(n_states):
            probabilities = [viterbi[s_prev, t - 1] * trans_prob[s_prev, s] * emis_prob[s, observation_sequence[t]] for s_prev in range(n_states)]
            viterbi[s, t] = max(probabilities)
            backpointer[s, t] = np.argmax(probabilities)

    # Termination step
    last_state = np.argmax(viterbi[:, -1])
    most_likely_sequence = [last_state]
    probability_of_sequence = viterbi[last_state, -1]

    # Traceback step
    for t in range(n_observations - 1, 0, -1):
        last_state = backpointer[last_state, t]
        most_likely_sequence.insert(0, last_state)

    return most_likely_sequence, probability_of_sequence

# Define the parameters of the HMM
states = ["/h/", "/e/", "/l/", "/o/"]
observations = ["O1", "O2", "O3", "O4"]
observation_map = {"O1": 0, "O2": 1, "O3": 2, "O4": 3}

start_probabilities = np.array([1.0, 0.0, 0.0, 0.0])
transition_probabilities = np.array([
    [0.0, 0.7, 0.3, 0.0],
    [0.0, 0.2, 0.6, 0.2],
    [0.0, 0.0, 0.3, 0.7],
    [0.0, 0.0, 0.1, 0.9]
])
emission_probabilities = np.array([
    [0.6, 0.2, 0.1, 0.1],
    [0.1, 0.7, 0.1, 0.1],
    [0.1, 0.1, 0.6, 0.2],
    [0.2, 0.1, 0.2, 0.5]
])

# Observation sequence captured by the system
observation_sequence = [observation_map[o] for o in ["O1", "O2", "O3", "O4"]]

# Run the Viterbi algorithm
most_likely_sequence, sequence_probability = viterbi_algorithm(
    observation_sequence, states, start_probabilities, transition_probabilities, emission_probabilities
)

# Convert state indices to phonemes
most_likely_states = [states[state] for state in most_likely_sequence]


In [2]:
# Output the results
print("Most likely sequence of phoneme states:", most_likely_states)
print("Probability of the most likely sequence:", sequence_probability)

# Inference
print("The decoded phoneme sequence matches the expected sequence (/h/, /e/, /l/, /o/) for the observation [O1, O2, O3, O4].")

Most likely sequence of phoneme states: ['/h/', '/e/', '/l/', '/o/']
Probability of the most likely sequence: 0.03704399999999999
The decoded phoneme sequence matches the expected sequence (/h/, /e/, /l/, /o/) for the observation [O1, O2, O3, O4].


#Inference

 The Viterbi algorithm has accurately identified the phoneme sequence (/h/, /e/, /l/, /o/) corresponding to the observation [O1, O2, O3, O4]. This result demonstrates the model's effectiveness in decoding the expected word "hello" with a sequence probability of 0.03704.